File: src/cpp/explore/topquestions/easy/array/两个数组的交集II.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @File    :   两个数组的交集II.cpp
 * @Time    :   2020/07/25 11:46:04
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :   https://leetcode-cn.com/explore/featured/card/top-interview-questions-easy/1/array/26/
 */
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            return intersect(nums2, nums1);
        }

        unordered_map<int, int> cnt;
        for (int n : nums1) {
            cnt[n]++;
        }

        vector<int> ans;
        for (int n : nums2) {
            if (cnt[n] > 0) {
                ans.push_back(n);
                cnt[n]--;
            }
        }

        return ans;
    }
};

File: src/python/explore/topquestions/easy/array/两个数组的交集II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   两个数组的交集II.py
@Time    :   2020/07/26 10:06:25
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2):
            return self.intersect(nums2, nums1)

        cnt = {}
        for e in nums1:
            cnt[e] = cnt.get(e, 0) + 1

        ans = []
        for e in nums2:
            if e in cnt and cnt[e] > 0:
                ans.append(e)
                cnt[e] -= 1

        return ans
